connmgr/connmanager.go

```diff
@@ -0,0 +1,235 @@
+const (
+	ConnPending      ConnState = 0x0
+	ConnEstablished  ConnState = 0x1
+	ConnDisconnected ConnState = 0x2
+	ConnFailed       ConnState = 0x3
+)
+
+// ConnReq is the connection request to a network address. If permanent, the
+// connection will be retried on disconnection.
+type ConnReq struct {
+	Addr      string
+	Permanent bool
+
+	state      ConnState
+	retryCount uint32
+}
+
+// State is the connection state of the requested connection.
+func (c *ConnReq) State() ConnState {
+	return c.state
+}
+
+// ConnectedFunc is the signature of the callback function which is used to
+// subscribe to new connections.
+type ConnectedFunc func(*ConnReq, net.Conn)
+
+const (
+	// maxRetryDuration is the max amount of time retrying of a persistent
+	// connection is allowed to grow to.  This is necessary since the retry
+	// logic uses a backoff mechanism which increases the interval base done
+	// the number of retries that have been done.
+	maxRetryDuration = time.Minute * 5
+)
+
+// Config holds the configuration options related to the connection manager.
+type Config struct {
+	// MaxOutboundConnections is the maximum number of outbound network
+	// connections to maintain.
+	MaxOutboundConnections uint32
+
+	// RetryDuration is the duration to wait before retrying a permanent
+	// connection request.
+	RetryDuration time.Duration
+
+	// OnConnection is a callback that is fired when a new connection is established.
+	OnConnection ConnectedFunc
+
+	// GetNewAddress is a way to get an address to make a network connection to.
+	GetNewAddress AddressFunc
+
+	// Dial connects to the address on the named network.
+	Dial DialFunc
+}
+
+// handleConnected is used to queue a successful connection.
+type handleConnected struct {
+	c    *ConnReq
+	conn net.Conn
+}
+
+// handleDisconnected is used to remove a connection.
+type handleDisconnected struct {
+	c *ConnReq
+}
+
+// handleFailed is used to remove a pending connection.
+type handleFailed struct {
+	c   *ConnReq
+	err error
+}
+
+// ConnManager provides a manager to handle network connections.
+type ConnManager struct {
+	cfg      *Config
+	wg       sync.WaitGroup
+	conns    map[*ConnReq]net.Conn
+	requests chan interface{}
+	stop     int32
+	quit     chan struct{}
+}
+
+// handleFailedConn handles a connection failed due to a disconnect or any
+// other failure. If permanent, it retries the connection after the configured
+// retry duration. Otherwise, it makes a new connection request.
+func (cm *ConnManager) handleFailedConn(c *ConnReq) {
+	if atomic.LoadInt32(&cm.stop) != 0 {
+		return
+	}
+	if c.Permanent {
+		c.retryCount++
+		d := time.Duration(c.retryCount) * cm.cfg.RetryDuration
+		if d > maxRetryDuration {
+			d = maxRetryDuration
+		}
+		log.Debugf("Retrying connection to %v in: %v", c.Addr, d)
+		time.AfterFunc(d, func() {
+			cm.Connect(c)
+		})
+	} else if uint32(len(cm.conns)) < cm.cfg.MaxOutboundConnections {
+		go cm.Connect(&ConnReq{Addr: cm.cfg.GetNewAddress()})
```

Comment:
This looks to me like it will have a degenerate condition when the network adapter goes down (such as a wireless card turned off).  It will launch a new goroutine which will fail immediately due to the adapter being down and turn send a message back to the `connHandler` that it failed which will call this function and launch another one, etc.  

This would happen incredibly fast and almost positively outpace the goroutine schedule such that it would balloon the number of goroutines and goes memory usage to go nuts.

Have you tried this scenario?